File: app/schemas/referral.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List
from app.schemas.user import UserResponse
from app.schemas.transaction import TransactionBase

class ReferralTransactionBase(BaseModel):
    amount: float
    created_at: datetime

class ReferralTransactionCreate(ReferralTransactionBase):
    user_id: int
    referred_user_id: int

class ReferralTransactionResponse(ReferralTransactionBase):
    id: int
    user_id: int
    referred_user_id: int
    transaction_id: Optional[int] = None
    referred_user: UserResponse
    transaction: Optional[TransactionBase] = None

    class Config:
        from_attributes = True

class ReferralStats(BaseModel):
    total_referrals: int
    total_earnings: float
    recent_referrals: List[ReferralTransactionResponse]

class ReferralResponse(BaseModel):
    referral_code: str
    referral_url: str
    referral_earnings: float
    total_referrals: int
    referred_by: Optional[UserResponse] = None

class ReferralHistoryResponse(BaseModel):
    id: int
    user_id: int
    referred_user_id: int
    amount: float
    created_at: datetime
    transaction_id: Optional[int] = None
    referred_user: UserResponse
    transaction: Optional[TransactionBase] = None

    class Config:
        from_attributes = True

File: admin/src/api/referral.ts
import api from './config';

export interface ReferralData {
  referral_code: string;
  referral_url: string;
  referral_earnings: number;
  total_referrals: number;
  referred_by: {
    id: number;
    username: string;
    email: string;
  } | null;
}

export interface ReferralTransaction {
  id: number;
  amount: number;
  created_at: string;
  referred_user: {
    id: number;
    username: string;
    email: string;
  };
}

export interface ReferralStats {
  total_referrals: number;
  total_earnings: number;
  recent_referrals: ReferralTransaction[];
}

// Get current user's referral information
export const getMyReferral = async (): Promise<ReferralData> => {
  const response = await api.get<ReferralData>('/referral/my-referral');
  return response.data;
};

// Get referral statistics
export const getReferralStats = async (): Promise<ReferralStats> => {
  const response = await api.get<ReferralStats>('/referral/stats');
  return response.data;
};

// Get referral transactions
export const getReferralTransactions = async (params?: {
  skip?: number;
  limit?: number;
}): Promise<ReferralTransaction[]> => {
  const response = await api.get<ReferralTransaction[]>('/referral/transactions', { params });
  return response.data;
}; 